File: typescript/src/match.ts
/** Glob-style pattern matching utilities. */

/**
 * Match a value against a glob pattern.
 *
 * Supports:
 * - `*` matches everything
 * - `prefix*` matches strings starting with prefix
 * - `*suffix` matches strings ending with suffix
 * - `?` matches a single character
 * - Exact match when no wildcards are present
 */
export function globMatch(pattern: string, value: string): boolean {
  if (!pattern) return false;
  if (pattern === "*") return true;

  // Convert glob to regex
  const escaped = pattern.replace(/([.+^${}()|[\]\\])/g, "\\$1");
  const regexStr = "^" + escaped.replace(/\*/g, ".*").replace(/\?/g, ".") + "$";
  return new RegExp(regexStr).test(value);
}

/**
 * Return true if `patterns` is undefined/null (don't care) or any pattern matches `value`.
 */
export function listMatches(
  patterns: string[] | undefined,
  value: string,
): boolean {
  if (patterns === undefined || patterns === null) return true;
  return patterns.some((p) => globMatch(p, value));
}

File: python/src/agent_policy_guard/match.py
"""Glob-style pattern matching utilities."""

from __future__ import annotations

import fnmatch


def glob_match(pattern: str, value: str) -> bool:
    """Match a value against a glob pattern.

    Supports:
    - ``*`` matches everything
    - ``prefix*`` matches strings starting with prefix
    - ``*suffix`` matches strings ending with suffix
    - ``pre*suf`` matches strings starting with pre and ending with suf
    - Exact match when no wildcards are present

    Uses :func:`fnmatch.fnmatchcase` for full glob semantics (``*``,
    ``?``, ``[seq]``, ``[!seq]``).
    """
    if not pattern:
        return False
    if pattern == "*":
        return True
    return fnmatch.fnmatchcase(value, pattern)


def list_matches(patterns: list[str] | None, value: str) -> bool:
    """Return True if *patterns* is None (don't care) or any pattern matches *value*."""
    if patterns is None:
        return True
    return any(glob_match(p, value) for p in patterns)
